[cuda][cudacodec] Address build warnings

cudawarped opened the pull request:
`cudacodec` produces a number of build warnings, mainly introduced by myself :facepalm:.  

PR removes this pollution from the build output.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

asmorkalov commented on modules/cudacodec/src/video_reader.cpp:
```diff
@@ -345,14 +351,14 @@ namespace
         case VideoReaderProps::PROP_LRF_HAS_KEY_FRAME: {
-            const int iPacket = propertyVal - rawPacketsBaseIdx;
-            if (videoSource_->RawModeEnabled() && iPacket >= 0 && iPacket < rawPackets.size()) {
+            const int iPacket = static_cast<int>(propertyVal) - rawPacketsBaseIdx;
```
`propertyVal` is function output and could contain any value here. Looks like it's a bug.

cudawarped replied:
Its output and input, not sure what I was thinking here will alter this so that it makes more sense.

asmorkalov replied:
Please document "input" logic then.

cudawarped replied:
@asmorkalov I have added a new method `rawPackageHasKeyFrame()` to avoid the input output argument confusion in the `get()` method but now I am wondering if it is the right time to make some more alterations.  Specifically those below which I "wrongly" introduced to mirror the `VideoCapture` class:

- remove `grab()/retrieve()` for the decoded frame.  This combination is completely unecessary `grab()` doesn't save any processing, it only remove the need to pass a `GpuMat` when you don't want to retrieve the decoded frame.
- `propertyVal` of `double` in `get(const VideoReaderProps propertyId, CV_OUT double& propertyVal)` results in a lot of needless casting when the value will always be an integer.

What do you think?

cudawarped replied:
@asmorkalov I have pushed changes to demonstrate the removal of `grab()` and making `propertyVal` a `size_t` what do you think?

cudawarped commented on modules/cudacodec/src/video_source.hpp:
```diff
@@ -57,7 +57,7 @@ class VideoSource
 
     virtual FormatInfo format() const = 0;
     virtual void updateFormat(const FormatInfo& videoFormat) = 0;
-    virtual bool get(const int propertyId, double& propertyVal) const { return false; }
```
It was for `CuvidVideoSource` but I agree it will be much cleaner to move this to cuvid_video_source.h especially as `CUvideosource` is marked for removal in the next version of the Nvidia Video Codec SDK.

asmorkalov commented:
Tested manually with Ubuntu 18.04, CUDA 10.2 and GF 1080.

cudawarped commented on modules/cudacodec/src/NvEncoder.h:
```diff
@@ -252,7 +252,7 @@ class NvEncoder
     * @brief This function returns the completion event.
     */
-    void* GetCompletionEvent(uint32_t eventIdx) { return (m_vpCompletionEvent.size() == m_nEncoderBuffer) ? m_vpCompletionEvent[eventIdx] : nullptr; }
+    void* GetCompletionEvent(uint32_t eventIdx) { return (m_vpCompletionEvent.size() == static_cast<size_t>(m_nEncoderBuffer)) ? m_vpCompletionEvent[eventIdx] : nullptr; }
```
We could but it may cause underflow errors without careful checking?

asmorkalov replied:
The value is initialized at NvEncoder.cpp:220, but is not checked right now. I propose to migrate to size_t and add a check there.

asmorkalov replied:
ok! Agreed!

asmorkalov commented on modules/cudacodec/include/opencv2/cudacodec.hpp:
```diff
@@ -380,59 +378,18 @@ class CV_EXPORTS_W VideoReader
     */
     CV_WRAP virtual FormatInfo format() const = 0;
 
-    /** @brief Grabs the next frame from the video source.
-
-    @param stream Stream for the asynchronous version.
-    @return `true` (non-zero) in the case of success.
-
-    The method/function grabs the next frame from video file or camera and returns true (non-zero) in
-    the case of success.
-
-    The primary use of the function is for reading both the encoded and decoded video data when rawMode is enabled.  With rawMode enabled
-    retrieve() can be called following grab() to retrieve all the data associated with the current video source since the last call to grab() or the creation of the VideoReader.
-     */
-    CV_WRAP virtual bool grab(Stream& stream = Stream::Null()) = 0;
```
Why do you removed grab? It breaks compatibility and it's still mentioned in other methods.

cudawarped replied:
> now I am wondering if it is the right time to make some more alterations. Specifically those below which I "wrongly" introduced to mirror the VideoCapture class:
> 
>    - remove grab()/retrieve() for the decoded frame. This combination is completely unecessary grab() doesn't save any processing, it only remove the need to pass a GpuMat when you don't want to retrieve the decoded frame.
>    -  propertyVal of double in get(const VideoReaderProps propertyId, CV_OUT double& propertyVal) results in a lot of needless casting when the value will always be an integer.
> 
> What do you think?

I removed `grab()` in that commit so you could take a look.  It was an unecessary addition which adds a lot of code mess to `VideoCapture` and was "wrongly" introduced to mirror `VideoCapture` without actually mirroring the funcitonality.  I can back it out and place it in another PR if you think agree with its removal?

cudawarped replied:
@asmorkalov Added `grab()` back into to this PR.

cudawarped commented:
@asmorkalov Updated PR:
- Rebased onto 4.x
- Addressed new warnings
- Updated the python test which was broken to check PR doesn't break bindings
- Kept the new method `rawPackageHasKeyFrame()` to avoid the input output argument confusion in the `get()` method

Can you review when you have chance please?

asmorkalov commented on modules/cudacodec/src/video_decoder.cpp:
```diff
@@ -76,7 +76,7 @@ static const char* GetVideoCodecString(cudaVideoCodec eCodec) {
     }
     for (int i = cudaVideoCodec_NumCodecs + 1; i < sizeof(aCodecName) / sizeof(aCodecName[0]); i++) {
         if (eCodec == aCodecName[i].eCodec) {
-            return aCodecName[eCodec].name;
```
Need to check.
